_shared/project/conf/supervisord-dev.conf

```diff
@@ -1,27 +1,37 @@
 [supervisord]
 nodaemon = true
 silent = true
-{% if cookiecutter.get("_directory") == "pyapp" %}
+{% if include_exists("conf/supervisord-dev.conf.json") %}
+{% for program_name, program_settings in include_json("conf/supervisord-dev.conf.json").get("programs", {}).items() %}
 
-[program:app]
-command=python3 -m {{ cookiecutter.package_name }}
-stdout_events_enabled=true
-stderr_events_enabled=true
-stopsignal = KILL
-stopasgroup = true
-{% endif %}
-{% if cookiecutter.get("_directory") == "pyramid-app" %}
+[program:{{ program_name }}]
+command={{ program_settings.pop("command") }}
+stdout_events_enabled={{ program_settings.pop("stdout_events_enabled", "true") }}
+stderr_events_enabled={{ program_settings.pop("stderr_events_enabled", "true") }}
+stopsignal={{ program_settings.pop("stopsignal", "KILL") }}
+stopasgroup={{ program_settings.pop("stopasgroup", "true") }}
+{% for setting_name, setting_value in program_settings.items() %}
+{{ setting_name }}={{ setting_value }}
+{% endfor %}
+{% endfor %}
+{% else %}
+{% if cookiecutter._directory == "pyramid-app" %}
 
 [program:web]
 command=gunicorn --bind :{{ cookiecutter.port }} --workers 1 --reload --timeout 0 --paste conf/development.ini
 stdout_events_enabled=true
 stderr_events_enabled=true
-stopsignal = KILL
-stopasgroup = true
-{% endif %}
-{% if include_exists("conf/supervisord-dev.conf") %}
+stopsignal=KILL
+stopasgroup=true
+{% else %}
 
-{{ include("conf/supervisord-dev.conf") -}}
+[program:app]
+command=python3 -m {{ cookiecutter.package_name }}
+stdout_events_enabled=true
+stderr_events_enabled=true
+stopsignal=KILL
+stopasgroup=true
+{% endif %}
```

Comment:
:+1: Moved this into a macro:

* https://github.com/hypothesis/cookiecutters/blob/f3e93dc640da1198d146dcdb89bc019bc64a56dd/_shared/project/conf/supervisord-dev.conf#L1-L27
* https://github.com/hypothesis/cookiecutters/blob/f3e93dc640da1198d146dcdb89bc019bc64a56dd/_shared/project/conf/supervisord.conf#L1-L29

It's a shame that I have to write it twice. But I think :crossed_fingers: this situation is unique to the twin `supervisord[-dev].conf` files: two files that're almost but not exactly the same. Hoping that problem won't come up again, in which case it's probably not worth trying to figure out a custom solution to the repetition (since we can't use normal Jinja2 includes or imports).

Reply:
> Could you include a default `conf/supervisord-dev.conf.json` for the various project types as a templated out piece of data instead of having two mechanisms?

I don't think this would work, no. When creating a new project we don't have any way to control what order cookiecutter renders the files in, so we wouldn't be able to ensure that `.cookiecutter/includes/supervisord-dev.conf.json` gets rendered before `conf/supervisord-dev.conf` does. There _might_ be something we can do with a cookiecutter pre-gen hook here?? But that feels like quite a costly hack.

Generally I think templates need to be written to work in the absence of include files from the project for two cases:

1. When generating a new project none of the include files will exist yet
2. When updating an existing project any include files might be missing from the new project. For example an include file might have been deleted from the project. Or perhaps a new include got added to the templates since the last time this project was updated

I think any hacks that we try to get around this reality are gonna end up pretty messy